src/properties/rich_club.c

```diff
@@ -0,0 +1,116 @@
+    igraph_integer_t size = igraph_vector_int_size(vertex_order);
+    IGRAPH_CHECK(igraph_vector_int_resize(map, size)); // resize map
+
+    for (igraph_integer_t i = 0; i < size; i++) {
+        VECTOR(*map)[VECTOR(*vertex_order)[i]] = i;
+    }
+    return IGRAPH_SUCCESS;
+}
+
+igraph_error_t igraph_rich_club_density_sequence(const igraph_t *graph,
+                                                 const igraph_vector_int_t *vertex_order,
+                                                 igraph_bool_t directed,
+                                                 igraph_bool_t loops,
+                                                 igraph_vector_t *res) {
+    /* TODO:
+     * implement directed vs. undirected functionality
+     */
+    igraph_integer_t numVertices = igraph_vcount(graph);
+
+    // Error: vertex_order wrong size
+    if (igraph_vector_int_size(vertex_order) != numVertices) {
+        IGRAPH_ERROR("Invalid vertex order length.", IGRAPH_EINVAL);
+    }
+
+    igraph_integer_t numEdges = igraph_ecount(graph);
+    igraph_vector_int_t edges;
+    igraph_vector_int_t edgesRemainingAfter;
+
+    IGRAPH_CHECK(igraph_vector_int_init(&edges, 0));
+    IGRAPH_CHECK(igraph_vector_int_init(&edgesRemainingAfter, numVertices));
+
+    IGRAPH_CHECK(igraph_vector_resize(res, numVertices)); // resize res
+
+    IGRAPH_CHECK(igraph_get_edgelist(graph, &edges, 0)); // get list of edges
+
+    // get map of vertex ordering -> orderOf[id] = index of vertexID in vertex_order
+    igraph_vector_int_t orderOf;
+    IGRAPH_CHECK(igraph_vector_int_init(&orderOf, numVertices));
+    igraph_i_get_vertex_order_map(vertex_order, &orderOf);
+
+    for (igraph_integer_t i = 0; i < numEdges; i++) {   // O(E)
+        igraph_integer_t v1 = VECTOR(edges)[2 * i];     // endpoints
+        igraph_integer_t v2 = VECTOR(edges)[2 * i + 1];
+        igraph_integer_t orderV1 = VECTOR(orderOf)[v1]; // ordering of endpoints
+        igraph_integer_t orderV2 = VECTOR(orderOf)[v2];
+
+        igraph_integer_t edgeRemovalIndex = (orderV1 < orderV2 ? orderV1 : orderV2);
+
+        VECTOR(edgesRemainingAfter)[edgeRemovalIndex]++; // add to edge remaining count
+    }
+
+    // derive the final edgesRemainingAfter vector via accumulating from the end
+    igraph_integer_t total = 0;
+    for (igraph_integer_t i = numVertices - 1; i >= 0; i--) { // O(V)
+        total += VECTOR(edgesRemainingAfter)[i];
+        VECTOR(edgesRemainingAfter)[i] = total;
+    }
+
+    // density calculation
+    for (igraph_integer_t i = 0; i < numVertices; i++) { // O(V)
+        // (numVertices - i) = the number of vertices left on this loop
+        igraph_real_t totalPossibleEdges = (numVertices - i) * ((numVertices - i) - 1) / 2;
+        VECTOR(*res)[i] = VECTOR(edgesRemainingAfter)[i] / totalPossibleEdges;
+    }
+
+    igraph_vector_int_destroy(&edges);
+    igraph_vector_int_destroy(&orderOf);
+    igraph_vector_int_destroy(&edgesRemainingAfter);
```

Comment:
The code is in good shape so we can start working on error handling a bit.

Add here:

```
IGRAPH_FINALLY_CLEAN(3);
```

This removed the 3 destructors from the "finally" stack.

Please also order destructor calls in the opposite order of init calls. This is for readability and clarity.